### PR TITLE
Nv dma mapping fix

### PR DESCRIPTION
**Pull request by ferasd**
@haggaie 
@drossetti

### CODE REVIEW

File: nv_peer_mem.c [hunk 1]
```diff
@@ -138,12 +141,26 @@ static void nv_get_p2p_free_callback(void *data)
...
+#endif
+
 	/* For now don't set nv_mem_context->page_table to NULL, 
 	  * confirmed by NVIDIA that inflight put_pages with valid pointer will fail gracefully.
 	*/
 
 	ACCESS_ONCE(nv_mem_context->is_callback) = 1;
 	(*mem_invalidate_callback) (reg_handle, nv_mem_context->core_context);
+#if NV_DMA_MAPPING
+	ret = nvidia_p2p_free_dma_mapping(dma_mapping);
```
Comment (ferasd) [inline nv_peer_mem.c]: Do we need to unmap before the free ?

Comment (drossetti) [inline nv_peer_mem.c]: it works as nvidia_p2p_put_pages, dma_unmapping is implicitly done by NV driver after invalidation callback, but client code is in charge of freeing the dma_mapping obj.
so no, we don't need to call dma_unmap here.

File: nv_peer_mem.c [hunk 2]
```diff
@@ -251,6 +268,7 @@ static int nv_dma_map(struct sg_table *sg_head, void *context,
 			peer_err("error, incompatible dma mapping version 0x%08x\n",
 				 dma_mapping->version);
 			nvidia_p2p_dma_unmap_pages(pdev, page_table, dma_mapping);
+			nvidia_p2p_free_dma_mapping(dma_mapping);
```
Comment (drossetti) [inline nv_peer_mem.c]: nv-p2p.h is clear about that, nvidia_p2p_free_dma_mapping is not required. see below:

/*                                                                                                                                                                                                                 
 * @brief                                                                                                                                                                                                          
 *   Unmap the physical pages previously mapped to the third-party device by                                                                                                                                       
 *   nvidia_p2p_dma_map_pages().                                                                                                                                                                                   
 *                                                                                                                                                                                                                 
 ...                                                                                                                                                                                     
 *   The DMA mapping containing the DMA addresses used by the third-party                                                                                                                                          
 *   device, as retrieved with nvidia_p2p_dma_map_pages(). After this call                                                                                                                                         
 *   returns, neither this struct nor the addresses contained within will be                                                                                                                                       
 *   valid for use by the third-party device.                                                                                                                                                                      
 *                                                                                                                                                                                                                 
 * @return                                                                                                                                                                                                         
 *    0           upon successful completion.                                                                                                                                                                      
 *    -EINVAL     if an invalid argument was supplied.                                                                                                                                                             
 *    -EIO        if an unknown error occurred.                                                                                                                                                                    
 */
int nvidia_p2p_dma_unmap_pages(struct pci_dev *peer,
        struct nvidia_p2p_page_table *page_table,
        struct nvidia_p2p_dma_mapping *dma_mapping);

File: nv_peer_mem.c [hunk 3]
```diff
@@ -259,6 +277,7 @@ static int nv_dma_map(struct sg_table *sg_head, void *context,
 		ret = sg_alloc_table(sg_head, dma_mapping->entries, GFP_KERNEL);
 		if (ret) {
 			nvidia_p2p_dma_unmap_pages(pdev, page_table, dma_mapping);
+			nvidia_p2p_free_dma_mapping(dma_mapping);
```
Comment (drossetti) [inline nv_peer_mem.c]: same here

File: nv_peer_mem.c [hunk 4]
```diff
@@ -316,9 +335,11 @@ static int nv_dma_unmap(struct sg_table *sg_head, void *context,
 #if NV_DMA_MAPPING
 	{
 		struct pci_dev *pdev = to_pci_dev(dma_device);
-		if (nv_mem_context->dma_mapping)
+		if (nv_mem_context->dma_mapping) {
 			nvidia_p2p_dma_unmap_pages(pdev, nv_mem_context->page_table,
 						   nv_mem_context->dma_mapping);
+                        nvidia_p2p_free_dma_mapping(nv_mem_context->dma_mapping);
```
Comment (drossetti) [inline nv_peer_mem.c]: this is not needed as well

**Pull request merged**